homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,360 @@
+        """
+        return self._attr_in_progress
+
+    @property
+    def latest_version(self) -> str | None:
+        """Latest version available for install."""
+        return self._attr_latest_version
+
+    @property
+    def release_summary(self) -> str | None:
+        """Summary of the release notes or changelog.
+
+        This is not suitable for long changelogs, but merely suitable
+        for a short excerpt update description of max 255 characters.
+        """
+        return self._attr_release_summary
+
+    @property
+    def release_url(self) -> str | None:
+        """URL to the full release notes of the latest version available."""
+        return self._attr_release_url
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return self._attr_supported_features
+
+    @property
+    def title(self) -> str | None:
+        """Title of the software.
+
+        This helps to differentiate between the device or entity name
+        versus the title of the software installed.
+        """
+        return self._attr_title
+
+    @final
+    async def async_skip(self) -> None:
+        """Skip the current offered version to update."""
+        if (latest_version := self.latest_version) is None:
+            raise HomeAssistantError(f"Cannot skip an unknown version for {self.name}")
+        if self.current_version == latest_version:
+            raise HomeAssistantError(f"No update available to skip for {self.name}")
+        self.__skipped_version = latest_version
+        self.async_write_ha_state()
+
+    async def async_install(
+        self,
+        version: str | None = None,
+        backup: bool | None = None,
+    ) -> None:
+        """Install an update.
+
+        Version can be specified to install a specific version. When `None`, the
+        latest version needs to be installed.
+
+        The backup parameter indicates a backup should be taken before
+        installing the update.
+        """
+        await self.hass.async_add_executor_job(self.install, version, backup)
+
+    def install(
+        self,
+        version: str | None = None,
+        backup: bool | None = None,
+    ) -> None:
+        """Install an update.
+
+        Version can be specified to install a specific version. When `None`, the
+        latest version needs to be installed.
+
+        The backup parameter indicates a backup should be taken before
+        installing the update.
+        """
+        raise NotImplementedError()
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        if (current_version := self.current_version) is None or (
+            latest_version := self.latest_version
+        ) is None:
+            return None
+
+        if latest_version not in (current_version, self.__skipped_version):
+            return STATE_ON
+        return STATE_OFF
```

Comment:
Should update in progress also be a state? I guess it depends on what the update entity describes. If it describes the availability of an update, on/off seems fine. If it describes an update, maybe ready/applying/on_latest would fit better? After all, that is also what the entity row would show

Reply:
Good question, and initially I wrote it as such as well. I kept it binary and simple. The progress moved to an attribute.

My train of though here was two fold:

- it still isn't updated while updating.
- the current form allows for translations in the frontend.
- the main use case of the entity state would be dermining if there is an update pending (and automate notification on that). Less possible states makes that easier I guess.

Reply:
Ok 👍  It means we think of it as representing version.